png_cicp_chunk_proposal.md

```diff
@@ -13,31 +13,33 @@ An existing W3C group note, [BT2100-in-PNG]  specifies an approach which is limi
 * extensible signaling of color space based on H.273
 * does not require the presence of iCCP chunk and embedded ICC profiles
 
+## Non Requirements
+* compatibility with HDR colorspace definitions in any other raster image format (eg. JPEG, AVIF)
+
 ## Strawman approach
-[H.273](https://www.itu.int/rec/T-REC-H.273/en) specifies a controlled vocabulary for the parameterization of
-color space information.
+[H.273](https://www.itu.int/rec/T-REC-H.273/en) specifies a controlled vocabulary for the parameterization of color space information.
 
-Define a cICP chunk that contains the 7 bytes necessary to carry the
-H.273 color space parameters:
+Define a `cICP` chunk that contains the 7 bytes necessary to carry the H.273 color space parameters:
 
-* COLPRIMS, 2 bytes, One of the ColourPrimaries enumerated values specified in Rec. ITU-T H.273 | [ISO/IEC 23091-2]
-* TRANSFC, 2 bytes, One of the TransferCharacteristics enumerated values specified in Rec. ITU-T H.273 | [ISO/IEC 23091-2]
-* VIDFRNG, 1 byte, Value of the VideoFullRangeFlag specified in Rec. ITU-T H.273 | [ISO/IEC 23091-2]
+* **COLPRIMS**, 2 bytes, One of the ColourPrimaries enumerated values specified in Rec. ITU-T H.273 | [ISO/IEC 23091-2]
+* **TRANSFC**, 2 bytes, One of the TransferCharacteristics enumerated values specified in Rec. ITU-T H.273 | [ISO/IEC 23091-2]
+* **VIDFRNG**, 1 byte, Value of the VideoFullRangeFlag specified in Rec. ITU-T H.273 | [ISO/IEC 23091-2]
 
-[ed.: these are inspired from recent JPEG standards that incorporate
-H.273 color space parameters]
+[ed.: these are inspired from recent JPEG standards (eg. JPEG-XL) that incorporate H.273 color space parameters]
 
 [ed.: The MATOEFFS parameter is not included because PNG does not support YCbCr]
 
-The cICP chunk comes before IDAT chunk.
+NOTE: [ITU-T Series H Supplement 19](https://www.itu.int/rec/T-REC-H.Sup19-201910-I) summarize combinations of H.273 parameters corresponding to common baseband linear broadcasts and file-based Video-on-Demand(VOD) services.
+
+### Implementation Details
+The `cICP` chunk SHALL come before `IDAT` chunk.  
 
-When the cICP chunk is present, PNG decoders that recognize it shall ignore the following chunks:
-- iCCP
-- gAMA 
-- cHRM 
-- sRGB 
+A PNG SHALL not contain both a `cICP` chunk and an `iCCP` chunk. If a PNG decoder detects the presence of both a `cICP` and a `iCCP` chunk, the behavior is undefined.
```

Comment:
I don't think we should add this.

From the current PNG spec on the iCCP chunk [1]: "A PNG encoder that writes the iCCP chunk is encouraged to also write gAMA and cHRM chunks that approximate the ICC profile, to provide compatibility with applications that do not use the iCCP chunk. When the iCCP chunk is present, PNG decoders that recognize it and are capable of colour management [ICC] shall ignore the gAMA and cHRM chunks and use the iCCP chunk instead and interpret it according to [ICC-1] and [ICC-1A]. PNG decoders that are used in an environment that is incapable of full-fledged colour management should use the gAMA and cHRM chunks if present."

If a PNG cannot contain both chunks, encoders are forced to pick one without knowing what the decoder supports. Suppose my encoder writes the cICP chunk but the decoder only understands iCCP. Now I get neither.

Instead, we could follow that existing iCCP chunk specification and allow one of the chunks to be a fallback. And that is how it was already written.

[1] https://www.w3.org/TR/2003/REC-PNG-20031110/#11iCCP

Reply:
ICC is a specific color rendering (one of many)
CICP is a color space, not a color rendering, so it supports a large number of renderings.
Given this, it seems ICC should take priority, as it indicates that a specific rendering has been selected.

Reply:
I mostly follow your suggestion*. But I don't think we can do that. I think the chunk priorities have to match chronological order of chunk standardization.

If someone wanted to specify a color space (which is what the hardware and OSes support), they would have to remove the iCCP chunk. But that leaves a gap between now and when all decoders support cICP. During that time (forever?), the mismatch in decoder support means wider color gamuts become flaky.

Suppose browser A supports cICP and browser B doesn't. Both support iCCP. If the file contains iCCP then the cICP chunk would never be used and would be worthless. If the iCCP chunk is left out, browser A looks correct and browser B does not. So a web dev would likely leave the iCCP chunk in.

It would be difficult to know when it is safe to use only the cICP chunk. Even if browsers A, B, and C support it and that accounts for 99% of a site's users, a web dev would need to actively revisit and weigh trade offs. I fear this would severely hamper cICP support.


* As a browser developer, I'm constrained to what the OS and hardware support/provide. I am able to mark a texture as DCI-P3 in Vulkan but I cannot mark a texture as a specific color rendering, for example. It is easy for me to say "The file says it is in DCI-P3. I'll mark the texture as DCI-P3" and not consider other use cases.

I've been reading a lot to fill my knowledge gaps--ICC specs, wikipedia, etc. My understanding right now is that the color rendering represents the lighting of the scene and the colors that light could produce. IE if my light cannot produce red then a tomato might not show up in my picture. If my understanding is correct, that would make color rendering useful for editing. But for the case of just displaying the image, it isn't useful.

Maybe we could change the wording to something along the lines of "For display purposes, an iCCP chunk can be ignored when a cICP chunk is present." That would allow room for an editor to consider both chunks.

Perhaps on our call tonight you can help me if I misunderstand.

Reply:
Per 2021-02-24: permit cICP and iCCP and defer processor behavior.